internal/runtime/client/client_test.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	hooksv1alpha1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	hooksv1alpha3 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
+)
+
+func TestClient_httpCall(t *testing.T) {
+	g := NewWithT(t)
+
+	ctx := context.TODO()
+
+	discoveryHookHandler := func(w http.ResponseWriter, r *http.Request) {
+		response := &hooksv1alpha1.DiscoveryHookResponse{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "DiscoveryHookResponse",
+				APIVersion: hooksv1alpha1.GroupVersion.Identifier(),
+			},
+			Status: hooksv1alpha1.ResponseStatusSuccess,
+		}
+		respBody, err := json.Marshal(response)
+		if err != nil {
+			panic(err)
+		}
+		w.WriteHeader(http.StatusOK)
+		w.Write(respBody)
+	}
+
+	t.Run("pass all nil", func(t *testing.T) {
```

Comment:
Having a list of test cases [like this](https://github.dev/ykakarap/cluster-api/blob/52669e487530eb71e677d9b1b34e9a7a0287ae00/internal/runtime/catalog/test/catalog_test.go#L225-L268) for all the test we want to do?

This way they can share the setup code and the tests become more readable.